ampd/src/sui/verifier.rs

```diff
@@ -0,0 +1,194 @@
+use sui_json_rpc_types::{SuiEvent, SuiTransactionBlockResponse};
+use sui_types::base_types::{ObjectID, SuiAddress};
+
+use crate::handlers::sui_verify_msg::Message;
+use crate::types::Hash;
+
+const CONTRACT_CALL_EVENT: &str = "gateway::ContractCall";
+
+// TODO: update after Sui gateway event finalization
+#[derive(Deserialize)]
+struct ContractCall {
+    pub source_id: SuiAddress,
+    pub destination_chain: String,
+    pub destination_address: String,
+    pub payload_hash: Hash,
+}
+
+// event type is in the form of: <gateway_address>::gateway::ContractCall
+fn call_contract_type(gateway_address: &ObjectID) -> StructTag {
+    format!("{}::{}", gateway_address, CONTRACT_CALL_EVENT)
+        .as_str()
+        .parse()
+        .expect("failed to parse struct tag")
+}
+
+impl PartialEq<&Message> for &SuiEvent {
+    fn eq(&self, msg: &&Message) -> bool {
+        if self.type_ != call_contract_type(&self.package_id) {
+            return false;
+        }
+
+        serde_json::from_value::<ContractCall>(self.parsed_json.clone()).unwrap();
+
+        match serde_json::from_value::<ContractCall>(self.parsed_json.clone()) {
+            Ok(contract_call) => {
+                contract_call.source_id == msg.source_address
+                    && contract_call.destination_chain == msg.destination_chain.to_string()
+                    && contract_call.destination_address == msg.destination_address
+                    && contract_call.payload_hash == msg.payload_hash
+            }
+            _ => false,
+        }
+    }
+}
+
+fn get_event<'a>(
+    gateway_address: &SuiAddress,
+    transaction_block: &'a SuiTransactionBlockResponse,
+    event_seq: u64,
+) -> Option<&'a SuiEvent> {
+    if let Some(events) = transaction_block.events.as_ref() {
+        events.data.iter().find(|event| {
+            event.id.event_seq == event_seq
+                && <ObjectID as Into<SuiAddress>>::into(event.package_id) == *gateway_address
+        })
+    } else {
+        None
+    }
+}
+
+#[allow(dead_code)]
+pub fn verify_message(
+    gateway_address: &SuiAddress,
+    transaction_block: &SuiTransactionBlockResponse,
+    message: &Message,
+) -> bool {
+    match get_event(gateway_address, transaction_block, message.event_index) {
+        Some(event) => transaction_block.digest == message.tx_id && event == message,
+        None => false,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use ethers::abi::AbiEncode;
+    use move_core_types::language_storage::StructTag;
+    use sui_json_rpc_types::{SuiEvent, SuiTransactionBlockEvents, SuiTransactionBlockResponse};
+    use sui_types::{
+        base_types::{SuiAddress, TransactionDigest},
+        event::EventID,
+    };
+
+    use crate::handlers::sui_verify_msg::Message;
+    use crate::sui::verifier::verify_message;
+    use crate::types::{EVMAddress, Hash};
+
+    fn get_matching_msg_and_tx_block() -> (SuiAddress, SuiTransactionBlockResponse, Message) {
```

Comment:
generally I prefer to put test helper functions at the end, so you can directly see what the tests are